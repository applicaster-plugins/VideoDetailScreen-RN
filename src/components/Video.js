import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { not } from 'ramda';
import {
  View,
  Image,
  TouchableWithoutFeedback,
  NativeModules
} from 'react-native';
import { APVideoPlayer } from 'react-native-zapp-bridge';
import VideoBanner from './VideoBanner';
import assets from '../utils/assetsMap';
import Badge from './Badge';
import { getVideoDimensions, getCenterPosition, parameters } from '../utils';
import { styles } from './Video.styles';

class Video extends Component {
  constructor(props) {
    super(props);
    this.play = this.play.bind(this);
  }

  play() {
    const { playable } = this.props;
    const type = 'atom_entry';
    NativeModules.ZappPlayer.playFullScreen(type, playable, {});
  }

  renderPlayButton(imageSize) {
    const { isPlayable } = this.props;
    const playPosition = getCenterPosition(imageSize, {
      width: 30,
      height: 30
    });

    const playButtonImage = isPlayable ? (
      <Image source={assets.iconPlay} style={[styles.play, playPosition]} />
    ) : null;
    return playButtonImage;
  }

  renderBadge() {
    const { badgeText } = this.props;

    return badgeText ? (
      <Badge style={styles.label} labelText={badgeText} />
    ) : null;
  }

  getVideoTeaser() {
    const { image, isPlayable } = this.props;
    const imageSize = getVideoDimensions();
    const playButton = this.renderPlayButton(imageSize);

    return (
      <TouchableWithoutFeedback disabled={not(isPlayable)} onPress={this.play}>
        <View>
          <Image source={{ uri: image }} style={[styles.image, imageSize]} />
          <View style={[{ position: 'absolute', zIndex: 1 }, { ...imageSize }]}>
            {playButton}

            {this.renderBadge()}
          </View>
        </View>
      </TouchableWithoutFeedback>
    );
  }

  getVideoPlayer() {
    const { image: img, playable } = this.props;
    const imageSize = getVideoDimensions();
    return (
      <APVideoPlayer
        {...{ src: { type: 'atom_entry', model: playable }, img }}
        style={imageSize}
      />
    );
  }

  render() {
    const { startTS, isLive, showBanner, isPlayable } = this.props;
    const { inlinePlayer, hideLiveBar } = parameters;
    return (
      <View style={styles.container}>
        {isPlayable && inlinePlayer
          ? this.getVideoPlayer()
          : this.getVideoTeaser()}
        {showBanner && !hideLiveBar ? (
          <VideoBanner {...{ startTS, isLive }} />
        ) : null}
      </View>
    );
  }
}

Video.propTypes = {
  playable: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  image: PropTypes.string,
  startTS: PropTypes.number,
  isLive: PropTypes.bool,
  showBanner: PropTypes.bool,
  badgeText: PropTypes.string,
  isPlayable: PropTypes.bool
};

export default Video;
