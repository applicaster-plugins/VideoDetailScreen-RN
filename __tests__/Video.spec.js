import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Video from '../src/components/Video';

jest.mock('../src/utils/zappConfig');

const startTS = new Date('2019-05-01T00:00:00Z').getTime();

const playable = {
  type: {
    value: 'channel'
  },
  title: 'German Masters Series – TVM Open 2019 - Tag 3',
  id: 'a4d12b904e3a013739347054d2ab776f',
  summary: 'HEUTE / 09:00 Uhr',
  published: '',
  updated: '',
  author: {
    name: ''
  },
  media_group: [
    {
      type: 'image',
      media_item: [
        {
          src:
            'http://manager.vidibuscloud.net/mediafiles/8893/6575/6215/8eb65e3055be0136b7f5543d7edaf0a0/opener-tennis-allgemein_1024x576.jpg',
          key: 'image_base',
          type: 'image'
        }
      ]
    }
  ],
  link: {},
  extensions: {
    section: 'Tennis',
    start_time: '2019-05-09T09:00:00+00:00',
    end_time: '2019-05-09T11:00:00+00:00',
    open_with_plugin_id: 'SDVideoPage',
    is_live: true,
    body_text:
      '*Sportdeutschland.TV* zeigt die German Masters Series - TVM Open 2019 vom 07. bis zum 11.05. aus Troisdorf live und on demand.\n\nHol Dir SPORTDEUTSCHLAND.TV PLUS und genieße unser Angebot komplett werbefrei.'
  },
  content: {
    type: 'video/hls',
    src:
      'http://origin.vidibus.net/live/stream/itf-court-1.m3u8?realm=6e8cba30a0da0130491a6c626d58b44c&asset=966ef6104e3a01372994543d7edaf0a0_2&realm=6e8cba30a0da0130491a6c626d58b44c&asset=966ef6104e3a01372994543d7edaf0a0_2'
  }
};

it('Video renders correctly', () => {
  const tree = renderer
    .create(
      <Video
        playable={playable}
        image="http://manager.vidibuscloud.net/mediafiles/8893/6575/6215/8eb65e3055be0136b7f5543d7edaf0a0/opener-tennis-allgemein_1024x576.jpg"
        startTS={startTS}
        isLive={false}
        showBanner={true}
        badgeText="Test text"
        isPlayable={false}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
