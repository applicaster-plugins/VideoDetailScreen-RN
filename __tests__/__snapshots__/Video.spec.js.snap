// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video renders correctly 1`] = `
<View
  style={undefined}
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={undefined}
    testID={undefined}
  >
    <Image
      source={
        Object {
          "uri": "http://manager.vidibuscloud.net/mediafiles/8893/6575/6215/8eb65e3055be0136b7f5543d7edaf0a0/opener-tennis-allgemein_1024x576.jpg",
        }
      }
      style={
        Array [
          undefined,
          Object {
            "height": 232.875,
            "width": 414,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "position": "absolute",
            "zIndex": 1,
          },
          Object {
            "height": 232.875,
            "width": 414,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={undefined}
      >
        Test text
      </Text>
    </View>
  </View>
</View>
`;
